registry/consul/backend.go

```diff
@@ -36,25 +36,67 @@ func NewBackend(cfg *config.Consul) (registry.Backend, error) {
 	return &be{c: c, dc: dc, cfg: cfg}, nil
 }
 
-func (b *be) Register() error {
-	if !b.cfg.Register {
-		log.Printf("[INFO] consul: Not registering fabio in consul")
-		return nil
+func (b *be) Register(services []string) error {
+	if b.dereg == nil {
+		b.dereg = make(map[string](chan bool))
 	}
 
-	service, err := serviceRegistration(b.cfg)
-	if err != nil {
-		return err
+	// deregister unneeded service names
+	for serviceName := range b.dereg {
+		if !stringInSlice(serviceName, services) {
+			err := b.Deregister(serviceName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// register new service names
+	for _, serviceName := range services {
+		if _, ok := b.dereg[serviceName]; ok {
+			log.Printf("[DEBUG]: %q already registered (appears in dereg map)\n", serviceName)
+			continue
+		}
+
+		service, err := serviceRegistration(b.cfg, serviceName)
+		if err != nil {
+			return err
+		}
+
+		b.dereg[serviceName] = register(b.c, service)
 	}
 
-	b.dereg = register(b.c, service)
 	return nil
 }
 
-func (b *be) Deregister() error {
-	if b.dereg != nil {
-		b.dereg <- true // trigger deregistration
-		<-b.dereg       // wait for completion
+func (b *be) Deregister(serviceName string) error {
+	if len(serviceName) == 0 {
```

Comment:
Since `DeregisterAll()` exists I wouldn't use a magic value `""` to deregister all services. Every now and then I am tempted to do that but then opt for simplicity. `Deregister(service)` deregisters a service if it is currently registered, otherwise it does nothing. `DeregisterAll()` deregisters all currently registered services.